core/dbt/graph/graph.py

```diff
@@ -74,7 +74,22 @@ def get_subset_graph(self, selected: Iterable[UniqueId]) -> "Graph":
         new_graph = self.graph.copy()
         include_nodes = set(selected)
 
-        for node in self:
+        while True:
+
+            nodes_to_remove = list(
+                node
+                for node in new_graph
+                if node not in include_nodes and new_graph.degree(node) < 2
+            )
+            if not nodes_to_remove:
+                break
+            new_graph.remove_nodes_from(nodes_to_remove)
+
+        # sort remaining nodes by degree
+        remaining_nodes = list(new_graph.nodes())
+        remaining_nodes.sort(key=lambda node: new_graph.degree(node))
```

Comment:
This is an interesting heuristic. We agree it makes sense to consider nodes with smaller degree first for the reasons you outlined in the original issue. Would you consider evaluating a slightly different heuristic which we think might work as well or better?

The idea is that the number of edges (potentially) created by removing a node is the product of its in-degree and out-degree, and that perhaps degree is working because it is a rough proxy for that product. If that is the case, maybe sorting by the product instead would be even better?

```python
remaining_nodes.sort(key=lambda node: new_graph.in_degree(node) * new_graph.out_degree(node))
```

Reply:
Smart! Agreed.